packages/protocol-perf/src/index.ts

```diff
@@ -126,19 +133,54 @@ class DefaultPerfService implements Startable, PerfService {
       throw new Error('bytesToSendBack was not set')
     }
 
+    let lastAmountOfBytesSent = BigInt(0)
+    let lastReportedTime = Date.now()
+    let totalBytesSent = BigInt(0)
+
+    const initialStartTime = Date.now()
+
     await stream.sink(async function * () {
       while (bytesToSendBack > 0n) {
         let toSend: bigint = writeBlockSize
         if (toSend > bytesToSendBack) {
           toSend = bytesToSendBack
         }
+
         bytesToSendBack = bytesToSendBack - toSend
         yield uint8Buf.subarray(0, Number(toSend))
+
+        if (Date.now() - lastReportedTime > 1000) {
+          const output: PerfOutput = {
+            type: 'intermediary',
+            timeSeconds: (Date.now() - lastReportedTime) / 1000,
+            uploadBytes: BigInt(0).toString(),
+            downloadBytes: lastAmountOfBytesSent.toString()
+          }
+
+          // eslint-disable-next-line no-console
+          console.log(JSON.stringify(output))
+
+          lastReportedTime = Date.now()
+          lastAmountOfBytesSent = BigInt(0)
+        }
+
+        lastAmountOfBytesSent += toSend
+        totalBytesSent += toSend
       }
     }())
+
+    const finalOutput: PerfOutput = {
+      type: 'final',
+      timeSeconds: (Date.now() - initialStartTime) / 1000,
+      uploadBytes: BigInt(0).toString(),
+      downloadBytes: totalBytesSent.toString()
+    }
+
+    // eslint-disable-next-line no-console
+    console.log(JSON.stringify(finalOutput))
```

Comment:
Why are we printing this to console? Should we use a logger? should this be emitted as an event?

Reply:
Measurement output is printed to `stdout` , see https://github.com/libp2p/test-plans/tree/master/perf#adding-a-new-implementation-or-a-new-version for more details

Reply:
To add some more context, the test runner (i.e. the one starting the various implementation binaries) parses the test results off of a binaries stdout.

I can not judge whether `console.log` is the best way to print to stdout in JS.

Reply:
We can presumably buffer the output up and log it all at the end without affecting the test measurements?  As in, the runner just collects the results rather than measuring how long it takes the implementation to deliver the results?

In JS `console.log` is blocking so very slow.

Reply:
Not sure what you mean @achingbrain the test runner relies on the output (which to my knowledge the best way to be outputted in a cross-environment fashion is via `console.log`) in order to aggregate the various results. Where should the output be buffered to?

Reply:
`console.log` is a blocking operation that involves I/O, it's really slow and should not be used during timed sections of benchmarks.

Here is a simplistic example:

```js

function measureConsole () {
  const start = Date.now()

  for (let i = 0; i < 1_000_000; i++) {
    console.info('hello world')
  }

  return Date.now() - start
}

function measureStdOut () {
  const start = Date.now()

  for (let i = 0; i < 1_000_000; i++) {
    process.stdout.write('hello world\n')
  }

  return Date.now() - start
}

function measureBuffer () {
  const start = Date.now()
  const output = []

  for (let i = 0; i < 1_000_000; i++) {
    output.push('hello world')
  }

  const done = Date.now() - start

  console.info(JSON.stringify(output))

  return done
}

const cons = measureConsole()
const stdout = measureStdOut()
const buf = measureBuffer()

console.info('console.info', cons, 'ms')
console.info('process.stdout.write', stdout, 'ms')
console.info('buffered output', buf, 'ms')
```

Result:

```
console.info 4747 ms
process.stdout.write 2242 ms
buffered output 16 ms
```

> Where should the output be buffered to?

To a variable, nothing exiting.  See the `measureBuffer` function above.

The point is, buffer the measurement data, and write it to the console after the timed section of the test.

Reply:
Ok, so the perf README says:

> Every second the client must print the current progress to stdout

I've adjusted the timing sections so we aren't measuring anything while this happens.